File: frontend/src/api/development.ts
import type { DataResponse, ListResponse } from "../types";
import { http } from "./client";

export interface DevTask {
  id: string;
  code: string;
  title: string;
  description: string | null;
  status: string;
  branch_name: string | null;
  sandbox_path: string | null;
  repository_id: string | null;
  error: string | null;
  duration_ms: number | null;
  created_at: string | null;
  // present on detail
  plan?: DevPlan;
  changes?: DevChange[];
  tests?: DevTestRun[];
  review?: DevReview | null;
  pull_request?: DevPR | null;
  timeline?: DevStage[];
  metrics?: DevMetrics | null;
}

export interface DevPlan {
  summary: string;
  affected_files: string[];
  architecture_context: string;
  dependencies: string[];
  required_knowledge: string[];
  required_apis: string[];
  implementation_order: string[];
  risk_analysis: string;
  acceptance_criteria: string[];
}

export interface DevChange {
  id: string;
  path: string;
  change_type: string;
  language: string | null;
  content: string | null;
  diff: string | null;
  rationale: string | null;
  status: string;
}

export interface DevTestRun {
  kind: string;
  command: string | null;
  status: string;
  passed_count: number;
  failed_count: number;
  duration_ms: number | null;
}

export interface DevReviewComment {
  dimension: string;
  severity: string;
  file_path: string | null;
  message: string;
}

export interface DevReview {
  outcome: string;
  score: number;
  summary: string | null;
  comments: DevReviewComment[];
}

export interface DevPR {
  id: string;
  branch_name: string;
  base_branch: string;
  title: string;
  body: string | null;
  diff_summary: string | null;
  release_notes: string | null;
  status: string;
  commit_count: number;
  files_changed: number;
  additions: number;
  deletions: number;
}

export interface DevStage {
  stage: string;
  role: string | null;
  status: string;
  detail: string | null;
}

export interface DevMetrics {
  generated_files: number;
  files_changed: number;
  additions: number;
  deletions: number;
  commits: number;
  tests_run: number;
  tests_passed: number;
  review_score: number;
  duration_ms: number | null;
}

export interface DevFounderDash {
  running: number;
  completed: number;
  failed: number;
  pending_approvals: number;
  open_pull_requests: number;
  total_tasks: number;
  recent_tasks: DevTask[];
}

export const developmentApi = {
  tasks: (status?: string) =>
    http.get<ListResponse<DevTask>>(`/development/tasks${status ? `?status=${status}` : ""}`),
  task: (id: string) => http.get<DataResponse<DevTask>>(`/development/tasks/${id}`),
  createAndRun: (title: string, description?: string, provider_name?: string) =>
    http.post<DataResponse<DevTask>>("/development/run", { title, description, provider_name }),
  run: (id: string, provider_name?: string) =>
    http.post<DataResponse<DevTask>>(`/development/tasks/${id}/run`, { provider_name }),
  timeline: (id: string) => http.get<ListResponse<DevStage>>(`/development/tasks/${id}/timeline`),
  approve: (id: string, decision: string, notes?: string) =>
    http.post<DataResponse<DevTask>>(`/development/tasks/${id}/approve`, { decision, notes }),
  pendingApprovals: () => http.get<ListResponse<DevTask>>("/development/pending-approvals"),
  founderDashboard: () => http.get<DataResponse<DevFounderDash>>("/development/founder-dashboard"),
  team: () => http.get<ListResponse<AIAgent>>("/development/team"),
  orchestration: (id: string) =>
    http.get<DataResponse<Orchestration>>(`/development/tasks/${id}/orchestration`),
};

export interface AIAgent {
  employee: string;
  employee_code: string;
  role: string;
  authority: string;
  provider: string;
  responsibilities: string[];
  decision_rules: string[];
}

export interface Orchestration {
  stages: {
    stage: string;
    role: string | null;
    status: string;
    acting_employee: string | null;
    provider: string | null;
  }[];
  handoffs: {
    sequence: number;
    from_employee: string | null;
    to_employee: string | null;
    from_role: string | null;
    to_role: string | null;
    stage: string | null;
    summary: string | null;
  }[];
}

export const stageLabel = (s: string): string =>
  s.replace(/_/g, " ").replace(/\b\w/g, (c) => c.toUpperCase());

File: backend/app/models/development.py
"""Autonomous Software Development Engine ORM models (Sprint 13).

Models a software task's autonomous implementation: the task, its per-stage
sessions, the implementation plan, generated changes, code reviews and comments,
test runs, a pull-request draft, approval history, and implementation metrics.
Every change is explainable, reviewable, and reversible — and the Founder is the
final authority (no auto-merge, no auto-push).
"""

import uuid
from datetime import datetime

from sqlalchemy import DateTime, Float, ForeignKey, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.domain.development_enums import (
    ApprovalDecision,
    ChangeStatus,
    ChangeType,
    DevTaskStatus,
    PRStatus,
    ReviewDimension,
    ReviewOutcome,
    ReviewSeverity,
    SessionStatus,
    TestKind,
    TestStatus,
)
from app.models.base import GUID, Base, TimestampMixin, UUIDPrimaryKeyMixin


class DevelopmentTask(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "development_tasks"

    code: Mapped[str] = mapped_column(String(40), nullable=False, unique=True, index=True)
    title: Mapped[str] = mapped_column(String(300), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    status: Mapped[DevTaskStatus] = mapped_column(
        String(30), nullable=False, default=DevTaskStatus.QUEUED, index=True
    )
    work_item_id: Mapped[uuid.UUID | None] = mapped_column(
        GUID(), ForeignKey("work_items.id", ondelete="SET NULL"), nullable=True
    )
    repository_id: Mapped[uuid.UUID | None] = mapped_column(
        GUID(), ForeignKey("repositories.id", ondelete="SET NULL"), nullable=True
    )
    ai_employee_id: Mapped[uuid.UUID | None] = mapped_column(
        GUID(), ForeignKey("ai_employees.id", ondelete="SET NULL"), nullable=True
    )
    sandbox_path: Mapped[str | None] = mapped_column(String(1000), nullable=True)
    branch_name: Mapped[str | None] = mapped_column(String(200), nullable=True)
    # Optional existing-code modification intent (JSON): target_file, operation,
    # anchor/snippet/symbol/new_name, source_root. Absent => scaffold a new module.
    modification_spec: Mapped[str | None] = mapped_column(Text, nullable=True)
    error: Mapped[str | None] = mapped_column(Text, nullable=True)
    started_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    duration_ms: Mapped[int | None] = mapped_column(Integer, nullable=True)


class DevelopmentSession(UUIDPrimaryKeyMixin, Base):
    """One row per workflow stage — the task timeline."""

    __tablename__ = "development_sessions"

    task_id: Mapped[uuid.UUID] = mapped_column(
        GUID(), ForeignKey("development_tasks.id", ondelete="CASCADE"), nullable=False, index=True
    )
    stage: Mapped[str] = mapped_column(String(30), nullable=False)
    role: Mapped[str | None] = mapped_column(String(120), nullable=True)
    # WP7: the REAL AI employee that performed this stage + the provider selected
    # for them (mock until a live provider is configured).
    acting_ai_employee_id: Mapped[uuid.UUID | None] = mapped_column(
        GUID(), ForeignKey("ai_employees.id", ondelete="SET NULL"), nullable=True
    )
    provider_name: Mapped[str | None] = mapped_column(String(60), nullable=True)
    status: Mapped[SessionStatus] = mapped_column(
        String(20), nullable=False, default=SessionStatus.PENDING
    )
    detail: Mapped[str | None] = mapped_column(Text, nullable=True)
    sequence: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    started_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )


class ImplementationPlan(UUIDPrimaryKeyMixin, Base):
    __tablename__ = "implementation_plans"

    task_id: Mapped[uuid.UUID] = mapped_column(
        GUID(), ForeignKey("development_tasks.id", ondelete="CASCADE"), nullable=False, index=True
    )
    summary: Mapped[str | None] = mapped_column(Text, nullable=True)
    affected_files: Mapped[str | None] = mapped_column(Text, nullable=True)  # JSON
    architecture_context: Mapped[str | None] = mapped_column(Text, nullable=True)
    dependencies: Mapped[str | None] = mapped_column(Text, nullable=True)  # JSON
    required_knowledge: Mapped[str | None] = mapped_column(Text, nullable=True)  # JSON
    required_apis: Mapped[str | None] = mapped_column(Text, nullable=True)  # JSON
    implementation_order: Mapped[str | None] = mapped_column(Text, nullable=True)  # JSON
    risk_analysis: Mapped[str | None] = mapped_column(Text, nullable=True)
    acceptance_criteria: Mapped[str | None] = mapped_column(Text, nullable=True)  # JSON
    # Concrete, individually-verifiable requirements extracted from the task, and
    # the verification report produced before the PR gate (both JSON).
    requirements: Mapped[str | None] = mapped_column(Text, nullable=True)  # JSON
    verification: Mapped[str | None] = mapped_column(Text, nullable=True)  # JSON
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )


class GeneratedChange(UUIDPrimaryKeyMixin, Base):
    __tablename__ = "generated_changes"

    task_id: Mapped[uuid.UUID] = mapped_column(
        GUID(), ForeignKey("development_tasks.id", ondelete="CASCADE"), nullable=False, index=True
    )
    path: Mapped[str] = mapped_column(String(1000), nullable=False)
    old_path: Mapped[str | None] = mapped_column(String(1000), nullable=True)
    change_type: Mapped[ChangeType] = mapped_column(String(20), nullable=False)
    language: Mapped[str | None] = mapped_column(String(40), nullable=True)
    content: Mapped[str | None] = mapped_column(Text, nullable=True)
    diff: Mapped[str | None] = mapped_column(Text, nullable=True)
    rationale: Mapped[str | None] = mapped_column(Text, nullable=True)
    status: Mapped[ChangeStatus] = mapped_column(
        String(20), nullable=False, default=ChangeStatus.PROPOSED
    )
    sequence: Mapped[int] = mapped_column(Integer, nullable=False, default=0)


class CodeReview(UUIDPrimaryKeyMixin, Base):
    __tablename__ = "code_reviews"

    task_id: Mapped[uuid.UUID] = mapped_column(
        GUID(), ForeignKey("development_tasks.id", ondelete="CASCADE"), nullable=False, index=True
    )
    outcome: Mapped[ReviewOutcome] = mapped_column(String(30), nullable=False)
    score: Mapped[float] = mapped_column(Float, nullable=False, default=0.0)
    summary: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )


class ReviewComment(UUIDPrimaryKeyMixin, Base):
    __tablename__ = "review_comments"

    review_id: Mapped[uuid.UUID] = mapped_column(
        GUID(), ForeignKey("code_reviews.id", ondelete="CASCADE"), nullable=False, index=True
    )
    dimension: Mapped[ReviewDimension] = mapped_column(String(30), nullable=False)
    severity: Mapped[ReviewSeverity] = mapped_column(
        String(20), nullable=False, default=ReviewSeverity.INFO
    )
    file_path: Mapped[str | None] = mapped_column(String(1000), nullable=True)
    line: Mapped[int | None] = mapped_column(Integer, nullable=True)
    message: Mapped[str] = mapped_column(Text, nullable=False)


class TestRun(UUIDPrimaryKeyMixin, Base):
    __tablename__ = "test_runs"

    task_id: Mapped[uuid.UUID] = mapped_column(
        GUID(), ForeignKey("development_tasks.id", ondelete="CASCADE"), nullable=False, index=True
    )
    kind: Mapped[TestKind] = mapped_column(String(20), nullable=False)
    command: Mapped[str | None] = mapped_column(String(500), nullable=True)
    status: Mapped[TestStatus] = mapped_column(String(20), nullable=False)
    passed_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    failed_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    output: Mapped[str | None] = mapped_column(Text, nullable=True)
    duration_ms: Mapped[int | None] = mapped_column(Integer, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )


class PullRequest(UUIDPrimaryKeyMixin, Base):
    __tablename__ = "pull_requests"

    task_id: Mapped[uuid.UUID] = mapped_column(
        GUID(), ForeignKey("development_tasks.id", ondelete="CASCADE"), nullable=False, index=True
    )
    branch_name: Mapped[str] = mapped_column(String(200), nullable=False)
    base_branch: Mapped[str] = mapped_column(String(200), nullable=False, default="main")
    title: Mapped[str] = mapped_column(String(300), nullable=False)
    body: Mapped[str | None] = mapped_column(Text, nullable=True)
    diff_summary: Mapped[str | None] = mapped_column(Text, nullable=True)
    release_notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    status: Mapped[PRStatus] = mapped_column(String(20), nullable=False, default=PRStatus.DRAFT)
    commit_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    files_changed: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    additions: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    deletions: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )


class ApprovalHistory(UUIDPrimaryKeyMixin, Base):
    __tablename__ = "approval_history"

    task_id: Mapped[uuid.UUID] = mapped_column(
        GUID(), ForeignKey("development_tasks.id", ondelete="CASCADE"), nullable=False, index=True
    )
    pull_request_id: Mapped[uuid.UUID | None] = mapped_column(
        GUID(), ForeignKey("pull_requests.id", ondelete="SET NULL"), nullable=True
    )
    decision: Mapped[ApprovalDecision] = mapped_column(String(30), nullable=False)
    actor: Mapped[str | None] = mapped_column(String(200), nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )


class ImplementationMetrics(UUIDPrimaryKeyMixin, Base):
    __tablename__ = "implementation_metrics"

    task_id: Mapped[uuid.UUID] = mapped_column(
        GUID(), ForeignKey("development_tasks.id", ondelete="CASCADE"), nullable=False, index=True
    )
    generated_files: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    files_changed: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    additions: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    deletions: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    commits: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    tests_run: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    tests_passed: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    review_score: Mapped[float] = mapped_column(Float, nullable=False, default=0.0)
    duration_ms: Mapped[int | None] = mapped_column(Integer, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )


class DevelopmentHandoff(UUIDPrimaryKeyMixin, Base):
    """A recorded handoff of work between two AI employees (WP7)."""

    __tablename__ = "development_handoffs"

    task_id: Mapped[uuid.UUID] = mapped_column(
        GUID(), ForeignKey("development_tasks.id", ondelete="CASCADE"), nullable=False, index=True
    )
    sequence: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    from_employee_id: Mapped[uuid.UUID | None] = mapped_column(
        GUID(), ForeignKey("ai_employees.id", ondelete="SET NULL"), nullable=True
    )
    to_employee_id: Mapped[uuid.UUID | None] = mapped_column(
        GUID(), ForeignKey("ai_employees.id", ondelete="SET NULL"), nullable=True
    )
    from_role: Mapped[str | None] = mapped_column(String(160), nullable=True)
    to_role: Mapped[str | None] = mapped_column(String(160), nullable=True)
    stage: Mapped[str | None] = mapped_column(String(40), nullable=True)
    summary: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
